extensions/data-transfer/portability-data-transfer-generic/src/main/java/org/datatransferproject/datatransfer/generic/BlobbySerializer.java

```
@@ -106,72 +109,114 @@ public String getPath() {
 
-  static class BlobbyContainerPath {
-    private BlobbyStorageContainerResource container;
-    private String path;
+  static final String SCHEMA_SOURCE =
+      GenericTransferConstants.SCHEMA_SOURCE_BASE
+          + "/extensions/data-transfer/portability-data-transfer-generic/src/main/java/org/datatransferproject/datatransfer/generic/BlobbySerializer.java";
+  private static final String BLOB_ID_TO_NAME_KEY = "blobIdToNameKey";
+  private final JobStore jobStore;
+  private final Monitor monitor;
+  private BlobIdToName blobIdToName;
+
+  public BlobbySerializer(JobStore jobStore, Monitor monitor) {
+    this.jobStore = jobStore;
+    this.monitor = monitor;
+  }
+
+  private void addToJobStore(String id, String name) {
+    initialiseBlobIdToNameIfNot(JobMetadata.getJobId());
+    blobIdToName.add(id, name);
+  }
 
-    public BlobbyContainerPath(BlobbyStorageContainerResource container, String path) {
-      this.container = container;
-      this.path = path;
+  private void saveStateToStore() {
+    initialiseBlobIdToNameIfNot(JobMetadata.getJobId());
+    try {
+      jobStore.create(JobMetadata.getJobId(), BLOB_ID_TO_NAME_KEY, blobIdToName);
+    } catch (IOException e) {
+      monitor.severe(() -> "Unable to save blobIdToName map to JobStore", e);
+      throw new RuntimeException(e);
     }
+  }
 
-    public BlobbyStorageContainerResource getContainer() {
-      return container;
-    }
+  private String getFromStore(String id) {
+    initialiseBlobIdToNameIfNot(JobMetadata.getJobId());
+    return blobIdToName.get(id);
+  }
 
-    public String getPath() {
-      return path;
+  private void initialiseBlobIdToNameIfNot(UUID jobId) {
+    if (blobIdToName == null) {
+      try {
+        blobIdToName = jobStore.findData(jobId, BLOB_ID_TO_NAME_KEY, BlobIdToName.class);
+      } catch (IOException e) {
+        monitor.severe(() -> "Not able to find blobIdToNameKey map from JobStore", e);
+        throw new RuntimeException(e);
```

You could let the IOException propagate and use the Copier's retry logic to try again - might need to update the `serializer` interface

---

done